### DISCUSSION TITLE
npm caret clarification

### DISCUSSION THREAD

**Discussion by benomatis** [microsoft/rnx-kit#2266] [Q&A] [answered]
Maybe I'm misunderstanding something, but why would `rnx-align-deps` want to correct this while the [RN 0.71 profile](https://github.com/microsoft/rnx-kit/blob/main/packages/align-deps/src/presets/microsoft/react-native/profile-0.71.ts) should normally allow it?

- jest "^29.5.0" should be "^29.2.1"

Isn't caret supposed to allow minor version changes?

**Comment by tido64** [comment, marked as answer]
Hi, I think I answered the same question here: https://github.com/microsoft/rnx-kit/discussions/2086#discussioncomment-4612273

**Reply by benomatis** [reply]
Well, I guess then I have an inherent problem understanding how npm upgrades work. Why would I potentially end up with 2 versions? Why can I just not tell npm to upgrade (actually use the new version, and thus refer to it in `yarn.lock`), and to use the new version in package.json as well?

**Reply by tido64** [reply]
> Why would I potentially end up with 2 versions?

Say you added `"jest": "^29.2.1"` and Yarn resolves to a specific version, for example 29.2.2. Some time later, someone else adds `"jest": "^29.5.0"` in another package. When Yarn resolves this version now, it will get 29.5.0. Yarn doesn't update the previous resolution as that would break the promise of the lockfile, so it adds another entry. Now you have two versions of Jest.

> Why can I just not tell npm to upgrade (actually use the new version, and thus refer to it in yarn.lock), and to use the new version in package.json as well?

You can. As I mentioned in the other question, use [`yarn upgrade`](https://classic.yarnpkg.com/en/docs/cli/upgrade). You should also have a look at [yarn-deduplicate](https://github.com/scinos/yarn-deduplicate#yarn-deduplicate). They have more information there as well.

**Reply by benomatis** [reply]
but if I keep 29.2.1, and the new package requirement upgrades it to 29.5.0, I could still end up with a 2 versions of jest if a third package comes, say, with 29.3.0, so what is this solving, really?

`yarn upgrade` will not resolve this, it will just silence depandabot and my consciousness that I'm on the latest and greatest...

**Reply by tido64** [reply]
`^29.2.1` is a range. It's equivalent to telling package managers that any 29.x.x is good. You determine which version is installed by looking at your lockfile. You cannot determine this by just looking at `package.json`. `align-deps` ensures that your packages all declare the same range, thus ensuring that you have only a single entry of Jest for your packages in the lockfile. You can control what the package manager should resolve the version range to using `yarn upgrade`. If Dependabot is satisfied, it is because you ARE using the latest and greatest.

**Reply by kelset** [reply]
you can learn more here: https://github.com/npm/node-semver

**Reply by benomatis** [reply]
Thank you, guys, I guess I'll have to do some more reading, but it's not really anything specific to this package.
